### PR TITLE
feat: Read dependent pipelines for the run pipeline stage

### PR DESCRIPTION
**Pull request by ovidiupopa07**
in case there is a spinnaker pipeline with a run pipeline stage we go and fetch the dependent pipeline details before calling the migrator api (the dependent pipeline can be in the same application or in a different one - both use cases are supported)

Spinnaker input
```
[
   {
      "appConfig": {},
      "application": "ovidiu",
      "id": "daa1cd5f-c4c2-42f8-8f1d-c042eed99cde",
      "index": 0,
      "keepWaitingPipelines": false,
      "lastModifiedBy": "admin",
      "limitConcurrent": true,
      "name": "cronpipelinespinnaker",
      "schema": "1",
      "spelEvaluator": "v4",
      "stages": [
         {
            "name": "Wait",
            "refId": "1",
            "requisiteStageRefIds": [],
            "type": "wait",
            "waitTime": 10
         }
      ],
      "triggers": [
         {
            "cronExpression": "0 0/6 * 1/1 * ? *",
            "enabled": true,
            "id": "8cd49fcf-0dde-483e-8351-81e18b088d2e",
            "type": "cron"
         },
         {
            "cronExpression": "0 0/10 * 1/1 * ? *",
            "enabled": true,
            "id": "928c9166-daa1-4db0-be1d-0a2574e3cbf2",
            "type": "cron"
         }
      ],
      "updateTs": "1708935476000"
   },
   {
      "application": "ovidiu",
      "id": "af8dcd0f-5a4e-4c0b-9416-282465d6ed1b",
      "index": 1,
      "keepWaitingPipelines": false,
      "lastModifiedBy": "admin",
      "limitConcurrent": true,
      "name": "manualjudgement",
      "schema": "1",
      "spelEvaluator": "v4",
      "stages": [
         {
            "failPipeline": true,
            "instructions": "Hey dude",
            "judgmentInputs": [
               {
                  "value": "rollback"
               },
               {
                  "value": "proceed"
               }
            ],
            "name": "Manual Judgment",
            "notifications": [],
            "refId": "1",
            "requisiteStageRefIds": [],
            "type": "manualJudgment"
         }
      ],
      "triggers": [],
      "updateTs": "1709037251000"
   },
   {
      "application": "ovidiu",
      "id": "ab4bbbe0-5a24-4628-86f9-1023b34bfcca",
      "index": 2,
      "keepWaitingPipelines": false,
      "lastModifiedBy": "admin",
      "limitConcurrent": true,
      "name": "childPipeline",
      "parameterConfig": [
         {
            "default": "",
            "description": "",
            "hasOptions": false,
            "label": "",
            "name": "waitTime",
            "options": [
               {
                  "value": ""
               }
            ],
            "pinned": true,
            "required": true
         }
      ],
      "schema": "1",
      "spelEvaluator": "v4",
      "stages": [
         {
            "name": "Wait",
            "refId": "1",
            "requisiteStageRefIds": [],
            "type": "wait",
            "waitTime": "${parameters.waitTime}"
         }
      ],
      "triggers": [],
      "updateTs": "1709661707000"
   },
   {
      "application": "ovidiu",
      "id": "e12c8d35-1f73-48aa-86b4-3aee3c250193",
      "index": 3,
      "keepWaitingPipelines": false,
      "lastModifiedBy": "admin",
      "limitConcurrent": true,
      "name": "wait",
      "schema": "1",
      "spelEvaluator": "v4",
      "stages": [
         {
            "name": "Wait",
            "refId": "1",
            "requisiteStageRefIds": [],
            "type": "wait",
            "waitTime": 15
         }
      ],
      "triggers": [],
      "updateTs": "1709637228000"
   },
   {
      "application": "ovidiu",
      "id": "9d2bccb4-3a8b-45ba-b34d-805a6665f5f4",
      "index": 4,
      "keepWaitingPipelines": false,
      "lastModifiedBy": "admin",
      "limitConcurrent": true,
      "name": "parentpipeline2",
      "schema": "1",
      "spelEvaluator": "v4",
      "stages": [
         {
            "application": "ovidiu",
            "failPipeline": true,
            "name": "PipelineStage",
            "pipeline": "54cd0eb9-28c1-4c33-a819-df80ca1c014a",
            "pipelineParameters": {
               "waitTime": "10"
            },
            "refId": "1",
            "requisiteStageRefIds": [],
            "type": "pipeline",
            "waitForCompletion": true
         }
      ],
      "triggers": [],
      "updateTs": "1709711605000"
   },
   {
      "application": "ovidiu",
      "id": "54cd0eb9-28c1-4c33-a819-df80ca1c014a",
      "index": 5,
      "keepWaitingPipelines": false,
      "lastModifiedBy": "admin",
      "limitConcurrent": true,
      "name": "childpipeline1",
      "schema": "1",
      "spelEvaluator": "v4",
      "stages": [
         {
            "application": "ovidiu",
            "failPipeline": true,
            "name": "Pipeline",
            "pipeline": "9029f745-4b03-4559-9ee4-2f95e2bcfa82",
            "refId": "1",
            "requisiteStageRefIds": [],
            "type": "pipeline",
            "waitForCompletion": true
         }
      ],
      "triggers": [],
      "updateTs": "1709711594000"
   },
   {
      "application": "ovidiu",
      "id": "9029f745-4b03-4559-9ee4-2f95e2bcfa82",
      "index": 6,
      "keepWaitingPipelines": false,
      "lastModifiedBy": "admin",
      "limitConcurrent": true,
      "name": "childpipeline2",
      "schema": "1",
      "spelEvaluator": "v4",
      "stages": [
         {
            "name": "Wait",
            "refId": "1",
            "requisiteStageRefIds": [],
            "type": "wait",
            "waitTime": 15
         }
      ],
      "triggers": [],
      "updateTs": "1709711575000"
   }
]
```

Migrator input 
```
{
   "pipelines": [
      {
         "appConfig": {},
         "application": "ovidiu",
         "id": "daa1cd5f-c4c2-42f8-8f1d-c042eed99cde",
         "index": 0,
         "keepWaitingPipelines": false,
         "lastModifiedBy": "admin",
         "limitConcurrent": true,
         "name": "cronpipelinespinnaker",
         "schema": "1",
         "spelEvaluator": "v4",
         "stages": [
            {
               "name": "Wait",
               "refId": "1",
               "requisiteStageRefIds": [],
               "type": "wait",
               "waitTime": 10
            }
         ],
         "triggers": [
            {
               "cronExpression": "0 0/6 * 1/1 * ? *",
               "enabled": true,
               "id": "8cd49fcf-0dde-483e-8351-81e18b088d2e",
               "type": "cron"
            },
            {
               "cronExpression": "0 0/10 * 1/1 * ? *",
               "enabled": true,
               "id": "928c9166-daa1-4db0-be1d-0a2574e3cbf2",
               "type": "cron"
            }
         ],
         "updateTs": "1708935476000"
      },
      {
         "application": "ovidiu",
         "id": "af8dcd0f-5a4e-4c0b-9416-282465d6ed1b",
         "index": 1,
         "keepWaitingPipelines": false,
         "lastModifiedBy": "admin",
         "limitConcurrent": true,
         "name": "manualjudgement",
         "schema": "1",
         "spelEvaluator": "v4",
         "stages": [
            {
               "failPipeline": true,
               "instructions": "Hey dude",
               "judgmentInputs": [
                  {
                     "value": "rollback"
                  },
                  {
                     "value": "proceed"
                  }
               ],
               "name": "Manual Judgment",
               "notifications": [],
               "refId": "1",
               "requisiteStageRefIds": [],
               "type": "manualJudgment"
            }
         ],
         "triggers": [],
         "updateTs": "1709037251000"
      },
      {
         "application": "ovidiu",
         "id": "ab4bbbe0-5a24-4628-86f9-1023b34bfcca",
         "index": 2,
         "keepWaitingPipelines": false,
         "lastModifiedBy": "admin",
         "limitConcurrent": true,
         "name": "childPipeline",
         "parameterConfig": [
            {
               "default": "",
               "description": "",
               "hasOptions": false,
               "label": "",
               "name": "waitTime",
               "options": [
                  {
                     "value": ""
                  }
               ],
               "pinned": true,
               "required": true
            }
         ],
         "schema": "1",
         "spelEvaluator": "v4",
         "stages": [
            {
               "name": "Wait",
               "refId": "1",
               "requisiteStageRefIds": [],
               "type": "wait",
               "waitTime": "${parameters.waitTime}"
            }
         ],
         "triggers": [],
         "updateTs": "1709661707000"
      },
      {
         "application": "ovidiu",
         "id": "e12c8d35-1f73-48aa-86b4-3aee3c250193",
         "index": 3,
         "keepWaitingPipelines": false,
         "lastModifiedBy": "admin",
         "limitConcurrent": true,
         "name": "wait",
         "schema": "1",
         "spelEvaluator": "v4",
         "stages": [
            {
               "name": "Wait",
               "refId": "1",
               "requisiteStageRefIds": [],
               "type": "wait",
               "waitTime": 15
            }
         ],
         "triggers": [],
         "updateTs": "1709637228000"
      },
      {
         "application": "ovidiu",
         "id": "9d2bccb4-3a8b-45ba-b34d-805a6665f5f4",
         "index": 4,
         "keepWaitingPipelines": false,
         "lastModifiedBy": "admin",
         "limitConcurrent": true,
         "name": "parentpipeline2",
         "schema": "1",
         "spelEvaluator": "v4",
         "stages": [
            {
               "application": "ovidiu",
               "dependentPipeline": {
                  "application": "ovidiu",
                  "id": "54cd0eb9-28c1-4c33-a819-df80ca1c014a",
                  "index": 5,
                  "keepWaitingPipelines": false,
                  "lastModifiedBy": "admin",
                  "limitConcurrent": true,
                  "name": "childpipeline1",
                  "schema": "1",
                  "spelEvaluator": "v4",
                  "stages": [
                     {
                        "application": "ovidiu",
                        "dependentPipeline": {
                           "application": "ovidiu",
                           "id": "9029f745-4b03-4559-9ee4-2f95e2bcfa82",
                           "index": 6,
                           "keepWaitingPipelines": false,
                           "lastModifiedBy": "admin",
                           "limitConcurrent": true,
                           "name": "childpipeline2",
                           "schema": "1",
                           "spelEvaluator": "v4",
                           "stages": [
                              {
                                 "name": "Wait",
                                 "refId": "1",
                                 "requisiteStageRefIds": [],
                                 "type": "wait",
                                 "waitTime": 15
                              }
                           ],
                           "triggers": [],
                           "updateTs": "1709711575000"
                        },
                        "failPipeline": true,
                        "name": "Pipeline",
                        "pipeline": "9029f745-4b03-4559-9ee4-2f95e2bcfa82",
                        "refId": "1",
                        "requisiteStageRefIds": [],
                        "type": "pipeline",
                        "waitForCompletion": true
                     }
                  ],
                  "triggers": [],
                  "updateTs": "1709711594000"
               },
               "failPipeline": true,
               "name": "PipelineStage",
               "pipeline": "54cd0eb9-28c1-4c33-a819-df80ca1c014a",
               "pipelineParameters": {
                  "waitTime": "10"
               },
               "refId": "1",
               "requisiteStageRefIds": [],
               "type": "pipeline",
               "waitForCompletion": true
            }
         ],
         "triggers": [],
         "updateTs": "1709711605000"
      }
   ]
}

```
all other features are still working as expected

Spinnaker 
image
Harness
image

https://harness.atlassian.net/browse/ASSMT-98

### CODE REVIEW

File: pipelines.go [hunk 1]
```diff
@@ -106,6 +108,111 @@ func migrateSpinnakerPipelines() error {
...
+			pipelineId, okId := s["pipeline"].(string)
+
+			if okType && stageType == "pipeline" && okId {
+				// Operations on the filtered stage
+				fmt.Printf("Updated stage with pipeline ID %s\n", pipelineId)
+				//var dependentPipeline map[string]interface{}
+
+				pipelinesToRemove, err = addDependentPipelineRecursive(pipelines, s, pipelineId, pipelinesToRemove, err, authMethod)
+
+				//dependentPipeline, err = findPipelineById(authMethod, pipelineId)
```
Comment (kkotula) [inline pipelines.go:132]: Do we need this commented-out code? If so, let's keep it. Otherwise, let's remove that section. :)

Comment (ovidiupopa07) [reply pipelines.go]: already deleted it - i noticed the commented lines just now :)

File: pipelines.go [hunk 2]
```diff
@@ -237,8 +344,39 @@ func findPipelineIdByName(pipelines []PipelineDetails, name string) string {
...
+	return GetWithAuth(migrationReq.SpinnakerHost, "applications/"+appName+"/pipelineConfigs", authMethod, migrationReq.Auth64, migrationReq.Cert, migrationReq.Key, migrationReq.Insecure)
+}
+
+// this is because there's no endpoint in gate to fetch pipeline config based on a pipeline id
+func findPipelineById(authMethod string, appName string, pipelineId string) (map[string]interface{}, error) {
+
+	var jsonBody []byte
+	var err error
+	var pipelines []map[string]interface{}
+	jsonBody, err = getAllPipelines(authMethod, appName)
```
[the remark this answers is not in the document]
Comment (ovidiupopa07) [reply pipelines.go:341]: makes sense, yes

**Review by kkotula — APPROVED** [review]
🚀

**Pull request merged**